src/Workspaces/Core/Portable/Options/OptionService.cs

```diff
@@ -4,197 +4,236 @@
+                if (workspaceOptionSet == null)
+                {
+                    throw new ArgumentException(WorkspacesResources.OptionsDidNotComeFromWorkspace, paramName: nameof(optionSet));
+                }
+
+                var changedOptions = new List<OptionChangedEventArgs>();
+
+                lock (_gate)
+                {
+                    foreach (var optionKey in workspaceOptionSet.GetAccessedOptions())
                     {
-                        foreach (var serializer in optionSerializers)
+                        var setValue = optionSet.GetOption(optionKey);
+                        object currentValue = this.GetOption(optionKey);
+
+                        if (object.Equals(currentValue, setValue))
                         {
-                            // There can be options (ex, formatting) that only exist in only one specific language. In those cases,
-                            // feature's serializer should exist in only that language.
-                            if (!SupportedSerializer(optionKey, serializer.Metadata))
-                            {
-                                continue;
-                            }
+                            // Identical, so nothing is changing
+                            continue;
+                        }
+
+                        // The value is actually changing, so update
+                        changedOptions.Add(new OptionChangedEventArgs(optionKey, setValue));
 
-                            if (serializer.Value.TryPersist(optionKey, setValue))
+                        _currentValues = _currentValues.SetItem(optionKey, setValue);
+
+                        ImmutableArray<Lazy<IOptionSerializer, OptionSerializerMetadata>> optionSerializers;
+                        if (_featureNameToOptionSerializers.TryGetValue(optionKey.Option.Feature, out optionSerializers))
+                        {
+                            foreach (var serializer in optionSerializers)
                             {
-                                break;
+                                // There can be options (ex, formatting) that only exist in only one specific language. In those cases,
+                                // feature's serializer should exist in only that language.
+                                if (!SupportedSerializer(optionKey, serializer.Metadata))
+                                {
+                                    continue;
+                                }
+
+                                if (serializer.Value.TryPersist(optionKey, setValue))
+                                {
+                                    break;
+                                }
                             }
                         }
                     }
                 }
+
+                // Outside of the lock, raise the events on our task queue.
+                _taskQueue?.ScheduleTask(() =>
+                {
+                    RaiseEvents(changedOptions);
```

Comment:
so this now basically means, when option changed, effect now can happen sometime later since listener can be notified some time later since option change actually happened.

I am fine with it, just stating that since now some feature that require/need effect right away, might need some other mechanism than using option service.